ext/standard/basic_functions.c

```diff
@@ -1214,18 +1216,32 @@ PHP_FUNCTION(flush)
 /* {{{ Delay for a given number of seconds */
 PHP_FUNCTION(sleep)
 {
-	zend_long num;
+	double seconds;
 
 	ZEND_PARSE_PARAMETERS_START(1, 1)
-		Z_PARAM_LONG(num)
+		Z_PARAM_DOUBLE(seconds)
 	ZEND_PARSE_PARAMETERS_END();
 
-	if (num < 0) {
+	if (seconds < 0) {
 		zend_argument_value_error(1, "must be greater than or equal to 0");
 		RETURN_THROWS();
 	}
 
-	RETURN_LONG(php_sleep((unsigned int)num));
+#if HAVE_NANOSLEEP
+	struct timespec php_req, php_rem;
+
+	php_req.tv_sec = (time_t) ((uint64_t)(seconds * NANOS_IN_SEC) / NANOS_IN_SEC);
```

Comment:
1. Previously, sleeping for 0x7fff_ffff_ffff_ffff seconds was possible (not really useful and I'm fine with it being off by thousands of seconds since it's unlikely real applications will run that long)
2. This should emit an error if the value of seconds cannot be represented as a uint64_t. Currently, this is an integer overflow for whatever (uint64_t)(seconds * NANOS_IN_SEC)  gets cast to.
3. I think this should use https://en.cppreference.com/w/c/numeric/math/fmod instead to determine the value for nanoseconds(fmod(seconds, 1.0)) and multiply that by NANOS_IN_SECS, and use floor() instead to determine tv_sec.
4. I think this should also check https://en.cppreference.com/w/c/numeric/math/isfinite , e.g. for NAN

I'm in favor of the API change overall, and implementation issues are fixable.

Reply:
1. 2. 4. - now nan is rejected, but all positive values incl. infinity is supported, is it good so?
3. - done